### ISSUE TITLE
Using numpy functions in @guvectorize-d block

### ISSUE THREAD

**Issue by pbruneau** [numba/numba#2736]
If I issue:

    @guvectorize(['void(int32[:],int32[:],int32[:])'], '(n),()->(n)', nopython=True, target='cuda')
    def g(x,y,res):
        for i in range(x.shape[0]):
            res[i] = x[i] + y[0]
    
    res = np.zeros(4, dtype=np.int32)
    a = np.array([1, 2, 3, 4], dtype=np.int32)
    b = np.array([2], dtype=np.int32)
    g(a,b,res)
    print(res)

Everything runs fine, yielding `[3 4 5 6]` as expected. Supporting `target='cuda'` here was actually the point of issue https://github.com/numba/numba/issues/2579.

Now, say that I want to use a numpy function in my guvectorize-d code, as in the following (stupid) example:

    @guvectorize(['void(int32[:],int32[:],int32[:])'], '(n),()->(n)', nopython=True, target='cuda')
    def g(x,y,res):
        for i in range(x.shape[0]):
            res[i] = x[i] + y[0] + np.min(x)
    
    res = np.zeros(4, dtype=np.int32)
    a = np.array([1, 2, 3, 4], dtype=np.int32)
    b = np.array([2], dtype=np.int32)
    g(a,b,res)
    print(res)

And now I get: 

    Failed at nopython (nopython frontend)
    Unknown attribute 'min' of type Module(<module 'numpy' from '/usr/lib/python2.7/dist-packages/numpy/__init__.pyc'>)

Is this to be expected? I mean numpy primitives may not be used in GPU vectorized code? Noting that removing `target='cuda'`, the code snippet works.

**Comment by seibert** [comment]
Yes, this is to be expected.  Only elementwise or scalar operations are supported in GPU code at the moment.  We're adding some additional explanation of this to the documentation for the next Numba release.

**Comment by pbruneau** [comment]
Thank you for the answer,
but then considering a less stupid example, i.e. trying to vectorize matrix row sums operations:

    import numpy as np
    from pprint import pprint
    from numba import guvectorize, float32
    
    @guvectorize(['void(float32[:,:],float32[:])'], '(m,n)->(m)', nopython=True)
    def rowsums(x,res):
        for i in range(x.shape[0]):
            res[i] = np.sum(x[i,:])
    
    mat = np.array(np.random.normal(0, 1, 9).reshape((3,3)), dtype=np.float32)
    pprint(mat)
    res = np.zeros(3, dtype=np.float32)
    rowsums(mat, res)
    pprint(res)

Yields something like:

    array([[ 0.09680238,  0.3279545 ,  1.14452219],
           [-0.80764109, -0.85997015,  0.86730367],
           [ 0.75728172, -0.71831441,  0.05684401]], dtype=float32)
    array([ 1.56927907, -0.80030757,  0.09581132], dtype=float32)

As expected. Expected also is the crash (identical to reported above) when adding `target='cuda'`.
Do you mean that @guvectorize will never support such "row-operations" as shown in my post?

**Comment by seibert** [comment]
I would not say "never", but it is something we have not had time to implement.

Due to the way that threads work with CUDA, a straightforward translation of array operations into for-loops may not give the desired performance, so we need to consider that in the implementation.

**Comment by cristipurdel** [comment]
Based on this issue and [https://numba.pydata.org/numba-doc/dev/cuda/cudapysupported.html] there are no array functions currently supported.
Which is kind of difficult since I wanted to do smth like  matrix inverse 6x6 or matrix multiplication.
This basically means that we have to do all the array functions, low leve ?!?
From the link it says...

> Unsupported numpy features:
> 
>     array creation APIs.
>     array methods.
>     functions that returns a new array.
I would prefer that some like dot or inverse/transpose should be at least implemented since from my point of view they are pretty basic and usefull
